core/src/main/scala/kafka/coordinator/PidManager.scala

```diff
@@ -0,0 +1,155 @@
+    pidBlockInfo.append(",blockEndPID:" + blockEndPid + ")")
+    pidBlockInfo.toString()
+  }
+}
+
+class PidManager(val brokerId: Int,
+                 val zkUtils: ZkUtils) extends Logging {
+
+  this.logIdent = "[PID Manager " + brokerId + "]: "
+
+  private var currentPIDBlock: PidBlock = null
+  private var nextPID: Long = -1L
+
+  // grab the first block of PIDs
+  this synchronized {
+    getNewPidBlock()
+    nextPID = currentPIDBlock.blockStartPid
+  }
+
+  private def getNewPidBlock(): Unit = {
+    var zkWriteComplete: Boolean = false
+    while (!zkWriteComplete) {
+      // refresh current pid block from zookeeper again
+      val (dataOpt, zkVersion) = zkUtils.readDataAndVersionMaybeNull(ZkUtils.PidBlockPath)
+
+      // generate the new pid block
+      currentPIDBlock = dataOpt match {
+        case Some(data) =>
+          val currPIDBlock = PidManager.parsePidBlockData(data)
+          debug("Read current pid block %s, Zk path version %d".format(currPIDBlock, zkVersion))
+
+          if (currPIDBlock.blockEndPid > Long.MaxValue - PidManager.PidBlockSize) {
+            // we have exhausted all pids (wow!), treat it as a fatal error
+            fatal("Exhausted all pids as the next block's end pid is will has exceeded long type limit (current block end pid is %d)".format(currPIDBlock.blockEndPid))
+            throw new KafkaException("Have exhausted all pids.")
+          }
+
+          PidBlock(brokerId, currPIDBlock.blockEndPid + 1L, currPIDBlock.blockEndPid + PidManager.PidBlockSize)
+        case None =>
+          debug("There is no pid block yet (Zk path version %d), creating the first block".format(zkVersion))
+          PidBlock(brokerId, 0L, PidManager.PidBlockSize - 1)
+      }
+
+      val newPIDBlockData = PidManager.generatePidBlockJson(currentPIDBlock)
+
+      // try to write the new pid block into zookeeper
+      val (succeeded, version) = zkUtils.conditionalUpdatePersistentPath(ZkUtils.PidBlockPath, newPIDBlockData, zkVersion, Some(checkPidBlockZkData))
+      zkWriteComplete = succeeded
+
+      if (zkWriteComplete)
+        info("Acquired new pid block %s by writing to Zk with path version %d".format(currentPIDBlock, version))
+    }
+  }
+
+  private def checkPidBlockZkData(zkUtils: ZkUtils, path: String, expectedData: String): (Boolean, Int) = {
+    try {
+      val expectedPidBlock = PidManager.parsePidBlockData(expectedData)
+      val (dataOpt, zkVersion) = zkUtils.readDataAndVersionMaybeNull(ZkUtils.PidBlockPath)
+      val newPIDBlock = dataOpt match {
+        case Some(data) =>
+          val currPIDBlock = PidManager.parsePidBlockData(data)
+          if (currPIDBlock.equals(expectedPidBlock))
+            (true, zkVersion)
+          else
+            (false, zkVersion)
+        case None =>
+      }
+    } catch {
+      case e: Exception =>
+        warn("Error while checking for pid block Zk data on path %s: expected data %s".format(path, expectedData), e)
```

Comment:
nit: I think string interpolation is slightly preferred. Same for the rest of the logging.